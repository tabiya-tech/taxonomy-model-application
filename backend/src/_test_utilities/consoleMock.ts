/* eslint-disable @typescript-eslint/no-empty-function */
export {};
// Suppress chatty console during the tests
jest.spyOn(console, "error").mockImplementation(() => {});
jest.spyOn(console, "warn").mockImplementation(() => {});
jest.spyOn(console, "log").mockImplementation(() => {});
jest.spyOn(console, "info").mockImplementation(() => {});
