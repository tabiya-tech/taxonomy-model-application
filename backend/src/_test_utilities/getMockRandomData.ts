const sampleValidSpecialCharacters = "^°!\"§$%&/()=?`´*+#'-_.,;<>öäü@[]{}|¡“¶¢[]≠¿'„…∞~<•±æœ";
const sampleInternationCharacters = "αβγδεζηθικλμνξοπρστυφχψω";
const THREE_BYTE_UTF8_CHAR = "€";
export const WHITESPACE = " \n\r\t";

export function getTestString(length: number, prefix: string = ""): string {
  return (prefix + sampleValidSpecialCharacters + sampleInternationCharacters)
    .slice(0, length)
    .padEnd(length, THREE_BYTE_UTF8_CHAR);
}

export function get3ByteTestString(length: number, prefix: string = ""): string {
  return prefix + THREE_BYTE_UTF8_CHAR.repeat(length - prefix.length);
}

export function getRandomString(length: number) {
  let result = "";
  const characters = sampleValidSpecialCharacters + sampleInternationCharacters + WHITESPACE;
  const charactersLength = characters.length;
  let counter = 0;
  while (counter < length) {
    result += characters.charAt(Math.floor(Math.random() * charactersLength)); // NOSONAR
    counter += 1;
  }
  return result;
}

export function getRandomBoolean() {
  return Math.random() > 0.5; // NOSONAR
}

export function generateRandomUrl() {
  const letters = "abcdefghijklmnopqrstuvwxyz";
  const tlds = [".com", ".net", ".org", ".io"]; // List of top-level domains

  // Generate a random string of letters for the domain name
  let domain = "";
  for (let i = 0; i < 10; i++) {
    const randomIndex = Math.floor(Math.random() * letters.length); // NOSONAR
    domain += letters[randomIndex];
  }

  // Pick a random top-level domain from the list
  const randomTldIndex = Math.floor(Math.random() * tlds.length); // NOSONAR
  const tld = tlds[randomTldIndex];

  // Generate a random path with 1 to 3 segments
  const segments = Math.floor(Math.random() * 3) + 1; // NOSONAR
  let path = "";
  for (let i = 0; i < segments; i++) {
    const randomSegment = Math.random().toString(36).slice(2, 10); // NOSONAR
    path += "/" + randomSegment;
  }
  const protocols = ["", "https://", "http://www.", "http://", "http://www."]; // NOSONAR
  const protocol = protocols[Math.floor(Math.random() * protocols.length)]; // NOSONAR

  return protocol + domain + tld + path;
}

export function generateRandomDigitString(minDigits: number, maxDigits: number): string {
  let result = "";
  const characters = "0123456789";
  const digitCount = Math.floor(Math.random() * (maxDigits - minDigits + 1)) + minDigits; // NOSONAR
  const charactersLength = characters.length;

  for (let i = 0; i < digitCount; i++) {
    result += characters.charAt(Math.floor(Math.random() * charactersLength)); // NOSONAR
  }
  return result;
}

export function generateRandomAlphabeticalString(minDigits: number, maxDigits: number) {
  let result = "";
  const characters = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
  const digitCount = Math.floor(Math.random() * (maxDigits - minDigits + 1)) + minDigits; // NOSONAR
  const charactersLength = characters.length;

  for (let i = 0; i < digitCount; i++) {
    result += characters.charAt(Math.floor(Math.random() * charactersLength)); // NOSONAR
  }
  return result;
}

export function generateRandomNumber(min: number, max: number) {
  return Math.floor(Math.random() * (max - min + 1)) + min; // NOSONAR
}
