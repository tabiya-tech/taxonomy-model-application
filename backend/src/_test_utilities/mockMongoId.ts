import mongoose from "mongoose";

export function getMockStringId(index: number): string {
  return index.toString(16).padStart(24, "0");
}

export function getMockObjectId(index: number): mongoose.Types.ObjectId {
  return mongoose.Types.ObjectId.createFromHexString(getMockStringId(index));
}
