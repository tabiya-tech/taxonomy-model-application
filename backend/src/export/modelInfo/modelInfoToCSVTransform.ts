import { pipeline, Transform } from "stream";
import { getRepositoryRegistry } from "server/repositoryRegistry/repositoryRegistry";
import { stringify } from "csv-stringify";
import { IModelInfo } from "modelInfo/modelInfo.types";
import { Readable } from "node:stream";

const modelInfoHeaders = [
  "UUID",
  "Name",
  "Locale",
  "Description",
  "Version",
  "Released",
  "Previous",
  "ReleaseNotes",
  "OriginURI",
];

interface IModelInfoRow {
  UUID: string;
  Name: string;
  Locale: string;
  Description: string;
  Version: string;
  Released: "TRUE" | "FALSE";
  Previous: string;
  ReleaseNotes: string;
  OriginUUID: string;
}

export const transformModelInfoSpecToCSVRow = (modelInfo: IModelInfo): IModelInfoRow => {
  return {
    UUID: modelInfo.UUID,
    Name: modelInfo.name,
    Locale: modelInfo.locale.shortCode,
    Description: modelInfo.description,
    Version: modelInfo.version,
    Released: modelInfo.released ? "TRUE" : "FALSE",
    Previous: modelInfo.previousUUID,
    ReleaseNotes: modelInfo.releaseNotes,
    OriginUUID: modelInfo.originUUID,
  };
};

class ModelInfoToCSVRowTransformer extends Transform {
  constructor() {
    super({ objectMode: true });
  }

  _transform(
    modelInfo: IModelInfo,
    _encoding: BufferEncoding,
    callback: (error?: Error | null, data?: never) => void
  ): void {
    try {
      const transformedRow = transformModelInfoSpecToCSVRow(modelInfo);
      this.push(transformedRow);
      callback();
    } catch (cause: unknown) {
      // Make sure stringification doesn't fail, otherwise throwing an error will cause the stream to hang
      let json: string = "";
      try {
        json = JSON.stringify(modelInfo, null, 2);
      } finally {
        const error = new Error(`Failed to transform ModelInfo to CSV row: ${json}`);
        console.error(error, cause);
        callback(error);
      }
    }
  }
}

const ModelInfoToCSVTransform = async (modelId: string): Promise<Readable> => {
  // the stringify is a stream, and we need a new one every time we create a new pipeline
  const modelInfoStringifier = stringify({
    header: true,
    columns: modelInfoHeaders,
  });
  const modelInfo = await getRepositoryRegistry().modelInfo.getModelById(modelId);

  if (!modelInfo) {
    throw new Error("ModelInfo not found");
  }
  return pipeline(Readable.from([modelInfo]), new ModelInfoToCSVRowTransformer(), modelInfoStringifier, (cause) => {
    if (cause) {
      console.error(new Error("Transforming ModelInfo to CSV failed"), cause);
    }
  });
};

export default ModelInfoToCSVTransform;
