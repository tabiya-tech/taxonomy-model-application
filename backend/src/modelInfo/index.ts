import {APIGatewayProxyEvent} from "aws-lambda";
import {APIGatewayProxyResult} from "aws-lambda/trigger/api-gateway-proxy";
import {
  errorResponse,
  HTTP_VERBS,
  responseJSON,
  StatusCodes,
  STD_ERRORS_RESPONSES
} from "server/httpUtils";
import {IModelInfo, INewModelInfoSpec} from "./modelInfoModel";
import {getRepositoryRegistry} from "server/repositoryRegistry/repositoryRegisrty";
import {ajvInstance, ParseValidationError} from "../validator";
import {
  IModelInfoRequest,
  ModelInfoRequestSchema,
  ModelInfoResponseErrorCodes
} from 'api-specifications/modelInfo';

import {ErrorCodes} from "api-specifications/error";

import {ValidateFunction} from "ajv";
import {transform} from "./transform";
import {getResourcesBaseUrl} from "server/config/config";


export const handler: (event: APIGatewayProxyEvent/*, context: Context, callback: Callback*/)
  => Promise<APIGatewayProxyResult>
  = async (event: APIGatewayProxyEvent/*, context: Context, callback: Callback*/) => {
  //POST /modelInfo
  if (event?.httpMethod === HTTP_VERBS.POST) {
    return postModelInfo(event);
  }

  return STD_ERRORS_RESPONSES.METHOD_NOT_ALLOWED;
};

/**
 * @openapi
 *
 * /models:
 *     post:
 *       tags:
 *         - model
 *       requestBody:
 *         content:
 *           application/json:
 *             schema:
 *               $ref: '/components/schemas/modelInfoRequestSchema'
 *         required: true
 *       responses:
 *         '201':
 *           description: Successfully created the model
 *           content:
 *             application/json:
 *               schema:
 *                  $ref: '/components/schemas/modelInfoResponseSchema'
 *         '400':
 *           description: |
 *             Failed to process the  bad request.
 *             Further information can be found in the `message` of response body, which can have the following values:
 *              - `MODEL_COULD_NOT_VALIDATE`: One of the body parameter is invalid
 *              - `MALFORMED_BODY': body of request is malformed
 *           content:
 *             application/json:
 *               schema:
 *                 $ref: '/components/schemas/errorResponseSchema'
 *          '415':
 *           description: |
 *             Failed to process the request with an invalid content type.
 *             Further information can be found in the `message` of response body, which can have the following values:
 *              - `UNSUPPORTED_MEDIA_TYPE`: The content type  should application/json
 *           content:
 *             application/json:
 *               schema:
 *                 $ref: '/components/schemas/errorResponseSchema'
 *         '500':
 *           description: |
 *             Server failed to process the request.
 *             Further information can be found in the `message` of response body, which can have the following values:
 *              - `MODEL_COULD_NOT_BE_CREATED`: the server is unable to create the request
 *           content:
 *             application/json:
 *               schema:
 *                 $ref: '/components/schemas/errorResponseSchema'
 */
async function postModelInfo(event: APIGatewayProxyEvent) {
  if (!event.headers['Content-Type'] || !event.headers['Content-Type'].includes('application/json')) { // application/json;charset=UTF-8
    return errorResponse(StatusCodes.UNSUPPORTED_MEDIA_TYPE, ErrorCodes.UNSUPPORTED_MEDIA_TYPE, "Content-Type should be application/json", "Received Content-Type:" + event.headers['Content-Type']);
  }
  let payload: IModelInfoRequest;
  try {
    payload = JSON.parse(event.body as string);
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  } catch (error: any) {
    return errorResponse(StatusCodes.BAD_REQUEST, ErrorCodes.MALFORMED_BODY, "Payload is malformed, it should be a valid model json", error.message);
  }
  const validateFunction = ajvInstance.getSchema(ModelInfoRequestSchema.$id as string) as ValidateFunction;
  const isValid = validateFunction(payload);
  if (!isValid) {
    const errorDetail = ParseValidationError(validateFunction.errors);
    return errorResponse(StatusCodes.BAD_REQUEST, ModelInfoResponseErrorCodes.MODEL_COULD_NOT_VALIDATE, "Payload should conform to schema", errorDetail);
  }

  const newModelInfoSpec: INewModelInfoSpec = {
    name: payload.name,
    description: payload.description,
    locale: payload.locale
  };
  let newModelInfo: IModelInfo;
  try {
    newModelInfo = await getRepositoryRegistry().modelInfo.create(newModelInfoSpec);
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  } catch (error: any) {  //
    // Do not show the error message to the user as it can contain sensitive information such as DB connection string
    return errorResponse(StatusCodes.INTERNAL_SERVER_ERROR, ModelInfoResponseErrorCodes.DB_FAILED_TO_CREATE_MODEL, "Failed to create the model in the DB", "");
  }
  return responseJSON(StatusCodes.CREATED, transform(newModelInfo, getResourcesBaseUrl()));
}


