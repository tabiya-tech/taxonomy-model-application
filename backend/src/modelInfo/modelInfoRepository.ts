import mongoose from "mongoose";
import { randomUUID } from "crypto";
import { IModelInfo, IModelInfoDoc, INewModelInfoSpec } from "./modelInfo.types";
import { populateImportProcessStateOptions } from "./populateImportProcessStateOptions";
import { populateExportProcessStateOptions } from "./populateExportProcessStateOptions";

export interface IModelRepository {
  readonly Model: mongoose.Model<IModelInfoDoc>;

  /**
   * Creates a new ModelInfo entry.
   *
   * @param {INewModelInfoSpec} newModelSpec - The specification for the new ModelInfo entry.
   * @return {Promise<INewModelInfoSpec>} - A Promise that resolves to the newly created ModelInfo entry.
   * Rejects with an error if the ModelInfo entry cannot be created.
   */
  create(newModelSpec: INewModelInfoSpec): Promise<IModelInfo>;

  /**
   * Finds a ModelInfo entry by its ID.
   *
   * @param {string} modelId - The unique ID of the ModelInfo entry.
   * @return {Promise<IModelInfo|null>} - A Promise that resolves to the found ModelInfo entry or null if not found.
   * Rejects with an error if the operation fails.
   */
  getModelById(modelId: string): Promise<IModelInfo | null>;

  /**
   * Finds a ModelInfo entry by its UUID.
   *
   * @param {string} uuid - The unique ID of the ModelInfo entry.
   * @return {Promise<IModelInfo|null>} - A Promise that resolves to the found ModelInfo entry or null if not found.
   * Rejects with an error if the operation fails.
   */
  getModelByUUID(uuid: string): Promise<IModelInfo | null>;

  /**
   * Get all ModelInfo entries.
   *
   * @return {Promise<IModelInfo[]>} - A promise that resolves to an array with all the ModelInfo entries.
   * Rejects with an error if the operation fails.
   */
  getModels(): Promise<IModelInfo[]>;
}

export class ModelRepository implements IModelRepository {
  public readonly Model: mongoose.Model<IModelInfoDoc>;

  constructor(model: mongoose.Model<IModelInfoDoc>) {
    this.Model = model;
  }

  async create(newModelSpec: INewModelInfoSpec): Promise<IModelInfo> {
    try {
      //@ts-ignore
      if (newModelSpec.UUID !== undefined) {
        throw new Error("UUID should not be provided");
      }
      const newUUID = randomUUID();
      const newModelInfo = new this.Model({
        ...newModelSpec,
        UUID: newUUID,
        version: "",
        releaseNotes: "",
        released: false,
        importProcessState: new mongoose.Types.ObjectId(), // models are created empty and are populated asynchronously with data via an import process
      });
      newModelInfo.UUIDHistory.unshift(newUUID);
      await newModelInfo.save();
      await newModelInfo.populate([populateImportProcessStateOptions, populateExportProcessStateOptions]);
      return newModelInfo.toObject();
    } catch (e: unknown) {
      const err = new Error("ModelInfoRepository.create: create failed", { cause: e });
      console.error(err);
      throw err;
    }
  }

  async getModelById(modelId: string): Promise<IModelInfo | null> {
    try {
      const modelInfo = await this.Model.findById(modelId)
        .populate([populateImportProcessStateOptions, populateExportProcessStateOptions])
        .exec();
      return modelInfo != null ? modelInfo.toObject() : null;
    } catch (e: unknown) {
      const err = new Error("ModelInfoRepository.getModelById: getModelById failed", { cause: e });
      console.error(err);
      throw err;
    }
  }

  async getModelByUUID(modelUUID: string): Promise<IModelInfo | null> {
    try {
      // Without the $eq, the NoSQL injection can be prevented by setting the sanitizeFilter = true
      const filter = {
        UUID: { $eq: modelUUID },
      };
      const modelInfo = await this.Model.findOne(filter)
        .populate([populateImportProcessStateOptions, populateExportProcessStateOptions])
        .exec();
      return modelInfo != null ? modelInfo.toObject() : null;
    } catch (e: unknown) {
      const err = new Error("ModelInfoRepository.getModelByUUID: getModelByUUID failed", { cause: e });
      console.error(err);
      throw err;
    }
  }

  async getModels(): Promise<IModelInfo[]> {
    try {
      const modelInfos = await this.Model.find({})
        .populate([populateImportProcessStateOptions, populateExportProcessStateOptions])
        .exec();
      return modelInfos.map((modelInfo) => modelInfo.toObject());
    } catch (e: unknown) {
      const err = new Error("ModelInfoRepository.getModels: getModels failed", { cause: e });
      console.error(err);
      throw err;
    }
  }
}
