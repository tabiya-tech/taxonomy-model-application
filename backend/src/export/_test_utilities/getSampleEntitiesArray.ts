import { INewOccupationSpec, IOccupation } from "esco/occupation/occupation.types";
import { getMockRandomISCOGroupCode } from "_test_utilities/mockISCOCode";
import { getMockRandomOccupationCode } from "_test_utilities/mockOccupationCode";
import { ObjectTypes, OccupationType, RelationType } from "esco/common/objectTypes";
import { INewLocalizedOccupationSpec } from "esco/localizedOccupation/localizedOccupation.types";
import { INewISCOGroupSpec } from "esco/iscoGroup/ISCOGroup.types";
import { INewSkillSpec, ISkill, ReuseLevel, SkillType } from "esco/skill/skills.types";
import { INewSkillGroupSpec, ISkillGroup } from "esco/skillGroup/skillGroup.types";
import { getMockRandomSkillCode } from "_test_utilities/mockSkillGroupCode";
import { getMockStringId } from "_test_utilities/mockMongoId";
import { INewOccupationHierarchyPairSpec } from "esco/occupationHierarchy/occupationHierarchy.types";
import { INewSkillHierarchyPairSpec } from "esco/skillHierarchy/skillHierarchy.types";
import { INewOccupationToSkillPairSpec } from "esco/occupationToSkillRelation/occupationToSkillRelation.types";
import { INewSkillToSkillPairSpec } from "esco/skillToSkillRelation/skillToSkillRelation.types";

export const getSampleOccupationSpecs = (givenModelId: string, isLocal: boolean = false): INewOccupationSpec[] => {
  return Array.from<never, INewOccupationSpec>({ length: 100 }, (_, i) => ({
    ISCOGroupCode: getMockRandomISCOGroupCode(),
    definition: `definition_${i}`,
    regulatedProfessionNote: `regulatedProfessionNote_${i}`,
    scopeNote: `scopeNote_${i}`,
    altLabels: i % 2 ? [`altLabel_1`, `altLabel_2`] : [],
    code: getMockRandomOccupationCode(isLocal),
    preferredLabel: `Occupation_${i}`,
    modelId: givenModelId,
    originUUID: ``,
    ESCOUri: `ESCOUri_${i}`,
    description: `description_${i}`,
    importId: `importId_${i}`,
    occupationType: isLocal ? OccupationType.LOCAL : OccupationType.ESCO,
  }));
};
export const getSampleLocalizedOccupationSpecs = (
  givenEscoOccupations: IOccupation[]
): INewLocalizedOccupationSpec[] => {
  return givenEscoOccupations.map((occupation, i) => ({
    localizesOccupationId: occupation.id,
    altLabels: i % 2 ? [] : [`altLabel_1`, `altLabel_2`],
    modelId: occupation.modelId,
    description: `description_${i}`,
    importId: `importId_${i}`,
    occupationType: OccupationType.LOCALIZED,
  }));
};

export const getSampleISCOGroupSpecs = (givenModelId: string): INewISCOGroupSpec[] => {
  return Array.from<never, INewISCOGroupSpec>({ length: 100 }, (_, i) => ({
    code: getMockRandomISCOGroupCode(),
    preferredLabel: `ISCOGroup_${i}`,
    modelId: givenModelId,
    originUUID: ``,
    altLabels: i % 2 ? [`altLabel_1`, `altLabel_2`] : [],
    ESCOUri: `ESCOUri_${i}`,
    description: `description_${i}`,
    importId: `importId_${i}`,
  }));
};

export const getSampleSkillsSpecs = (givenModelId: string): INewSkillSpec[] => {
  return Array.from<never, INewSkillSpec>({ length: 100 }, (_, i) => ({
    preferredLabel: `Skill_${i}`,
    modelId: givenModelId,
    originUUID: ``,
    ESCOUri: `ESCOUri_${i}`,
    definition: `definition_${i}`,
    description: `description_${i}`,
    scopeNote: `scopeNote_${i}`,
    skillType: SkillType.Knowledge,
    reuseLevel: ReuseLevel.CrossSector,
    altLabels: i % 2 ? [`altLabel_1`, `altLabel_2`] : [],
    importId: `importId_${i}`,
  }));
};

export const getSampleSkillGroupsSpecs = (givenModelId: string): INewSkillGroupSpec[] => {
  return Array.from<never, INewSkillGroupSpec>({ length: 100 }, (_, i) => ({
    code: getMockRandomSkillCode(),
    preferredLabel: `SkillGroup_${i}`,
    modelId: givenModelId,
    originUUID: ``,
    ESCOUri: `ESCOUri_${i}`,
    description: `description_${i}`,
    scopeNote: `scopeNote_${i}`,
    altLabels: i % 2 ? [`altLabel_1`, `altLabel_2`] : [],
    importId: `importId_${i}`,
  }));
};

export const getSampleOccupationHierarchy = (givenModelId: string): INewOccupationHierarchyPairSpec[] => {
  return Array.from<never, INewOccupationHierarchyPairSpec>({ length: 100 }, (_, i) => ({
    parentId: getMockStringId(999),
    parentType: ObjectTypes.Occupation,
    childType: ObjectTypes.Occupation,
    childId: getMockStringId(i),
    modelId: givenModelId,
    importId: `importId_${i}`,
  }));
};

export const getSampleSkillsHierarchy = (
  givenSkills: ISkill[],
  givenSkillGroups: ISkillGroup[]
): INewSkillHierarchyPairSpec[] => {
  return givenSkills.map((skill) => ({
    parentId: givenSkillGroups[0].id,
    parentType: ObjectTypes.SkillGroup,
    childType: ObjectTypes.Skill,
    childId: skill.id,
    modelId: skill.modelId,
  }));
};

export const getSampleOccupationToSkillRelations = (
  givenOccupations: IOccupation[],
  givenSkills: ISkill[]
): INewOccupationToSkillPairSpec[] => {
  return givenSkills.map((skill, i) => ({
    requiringOccupationType: OccupationType.LOCAL,
    requiredSkillId: skill.id,
    requiringOccupationId: givenOccupations[i].id,
    relationType: RelationType.OPTIONAL,
  }));
};

export const getSampleSkillToSkillRelations = (givenSkills: ISkill[]): INewSkillToSkillPairSpec[] => {
  return givenSkills.map((skill, i) => ({
    requiringSkillId: givenSkills[i + 1]?.id ?? givenSkills[0].id,
    requiredSkillId: skill.id,
    relationType: RelationType.OPTIONAL,
  }));
};
