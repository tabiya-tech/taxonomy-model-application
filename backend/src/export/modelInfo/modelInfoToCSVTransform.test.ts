// Mute chatty console logs
import "_test_utilities/consoleMock";

import { getRepositoryRegistry } from "server/repositoryRegistry/repositoryRegistry";
import { getMockStringId } from "_test_utilities/mockMongoId";
import { getTestString } from "_test_utilities/getMockRandomData";
import ModelInfoToCSVTransform, * as SKillsToCSVTransformModule from "./modelInfoToCSVTransform";
import ImportProcessStateAPISpecs from "api-specifications/importProcessState";
import { parse } from "csv-parse/sync";
import { IModelRepository } from "modelInfo/modelInfoRepository";
import { IModelInfo } from "modelInfo/modelInfo.types";

const ModelInfoRepository = jest.spyOn(getRepositoryRegistry(), "modelInfo", "get");

const getMockModelInfo = (i: number): IModelInfo => {
  return {
    exportProcessState: [],
    importProcessState: {
      id: "",
      result: { errored: false, parsingErrors: false, parsingWarnings: false },
      status: ImportProcessStateAPISpecs.Enums.Status.PENDING,
      createdAt: new Date(),
      updatedAt: new Date(),
    },
    id: getMockStringId(i),
    UUID: `uuid_${i}`,
    UUIDHistory: [`uuid_1`, `uuid_2`],
    name: `name_${i}_${getTestString(10)}`,
    locale: {
      UUID: `localeUUID_${i}`,
      shortCode: `shortCode_${i}`,
      name: `name_${i}`,
    },
    description: `description_${i}_${getTestString(10)}`,
    license: `description_${i}_${getTestString(10)}`,
    version: `version_${i}`,
    released: true,
    releaseNotes: `releaseNotes_${i}_${getTestString(10)}`,
    createdAt: new Date(0), // use a fixed date to make the snapshot stable
    updatedAt: new Date(1), // use a fixed date to make the snapshot stable
  };
};

function setupModelInfoRepositoryMock(findByIdFn: () => IModelInfo | null) {
  const mockModelInfoRepository: IModelRepository = {
    Model: undefined as never,
    create: jest.fn().mockResolvedValue(null),
    getModelById: jest.fn().mockImplementationOnce(findByIdFn),
    getModels: jest.fn(),
    getModelByUUID: jest.fn(),
    getHistory: jest.fn(),
  };
  ModelInfoRepository.mockReturnValue(mockModelInfoRepository);
}

describe("ModelInfosDocToCsvTransform", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  test.each([
    ["released", true],
    ["unreleased", false],
  ])(
    "should correctly transform ModelInfo data to CSV when model is %s",
    async (_description: string, givenReleased: boolean) => {
      // GIVEN findAll returns a stream of ModelInfos
      const givenModelInfo = getMockModelInfo(2);
      givenModelInfo.released = givenReleased;
      setupModelInfoRepositoryMock(() => givenModelInfo);

      // WHEN the transformation is applied
      const transformedStream = await ModelInfoToCSVTransform(givenModelInfo.id);

      // THEN the output should be a stream
      const chunks = [];
      for await (const chunk of transformedStream) {
        chunks.push(chunk);
      }
      const actualCSVOutput = chunks.join("");

      // AND be a valid CSV
      const parsedObjects = parse(actualCSVOutput, { columns: true });
      // AND contain the occupation data
      expect(parsedObjects).toMatchSnapshot();
      expect(actualCSVOutput).toMatchSnapshot();

      // AND the stream should end
      expect(transformedStream.closed).toBe(true);
    }
  );

  test.each([
    ["is empty", []],
    ["has one item", [`uuid_1`]],
    ["has multiple items", [`uuid_1`, `uuid_2`]],
  ])(
    `should correctly transform ModelInfo data to CSV when UUIDHistory %s`,
    async (_description: string, givenUUIDHistory: string[]) => {
      // GIVEN findAll returns a stream of ModelInfos
      const givenModelInfo = getMockModelInfo(2);
      givenModelInfo.UUIDHistory = givenUUIDHistory;
      setupModelInfoRepositoryMock(() => givenModelInfo);

      // WHEN the transformation is applied
      const transformedStream = await ModelInfoToCSVTransform(givenModelInfo.id);

      // THEN the output should be a stream
      const chunks = [];
      for await (const chunk of transformedStream) {
        chunks.push(chunk);
      }
      const actualCSVOutput = chunks.join("");

      // AND be a valid CSV
      const parsedObjects = parse(actualCSVOutput, { columns: true });
      // AND contain the occupation data
      expect(parsedObjects).toMatchSnapshot();
      expect(actualCSVOutput).toMatchSnapshot();

      // AND the stream should end
      expect(transformedStream.closed).toBe(true);
    }
  );

  describe("should handle errors during stream processing", () => {
    test("should throw an error if no model by that id exists in the db", () => {
      // GIVEN that findByIdAndStream will return a stream with the ModelInfo
      setupModelInfoRepositoryMock(() => null);

      // WHEN the transformation is applied
      const transformedStream = ModelInfoToCSVTransform(getMockStringId(1));

      // THEN expect the given error to be thrown
      return expect(transformedStream).rejects.toThrowError("ModelInfo not found");
    });
    test("should log an error and end the stream when the transformModelInfoSpecToCSVRow throws", async () => {
      // GIVEN that findByIdAndStream will return a stream with the ModelInfo
      const givenModelInfo = getMockModelInfo(3);
      setupModelInfoRepositoryMock(() => givenModelInfo);
      // AND the transformModelInfoToCSVRow will throw an error
      const givenError = new Error("Mocked Transformation Error");
      const transformFunctionSpy = jest
        .spyOn(SKillsToCSVTransformModule, "transformModelInfoSpecToCSVRow")
        .mockImplementationOnce(() => {
          throw givenError;
        });

      // WHEN the transformation stream is consumed
      const transformedStream = await ModelInfoToCSVTransform(givenModelInfo.id);

      // THEN expect the given error to be thrown
      await expect(async () => {
        //  iterate to consume the stream
        for await (const _ of transformedStream) {
          // do nothing
        }
      }).rejects.toThrowError("Failed to transform ModelInfo to CSV row");
      // AND the error to be logged
      const expectedLoggedItem = JSON.stringify(transformFunctionSpy.mock.calls[0][0], null, 2);
      const expectedErrorMessage = `Failed to transform ModelInfo to CSV row: ${expectedLoggedItem}`;

      expect(console.error).toHaveBeenCalledWith(
        expect.toMatchErrorWithCause(expectedErrorMessage, givenError.message)
      );
      // AND the stream to end
      expect(transformedStream.closed).toBe(true);
    });
  });
});
