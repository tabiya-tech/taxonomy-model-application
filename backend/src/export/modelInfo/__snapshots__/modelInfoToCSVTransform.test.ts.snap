// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModelInfosDocToCsvTransform should correctly transform ModelInfo data to CSV 1`] = `
[
  {
    "Description": "description_2_^°!"§$%&/(",
    "Locale": "shortCode_2",
    "Name": "name_2_^°!"§$%&/(",
    "OriginURI": "",
    "Previous": "previousUUID_2",
    "ReleaseNotes": "releaseNotes_2_^°!"§$%&/(",
    "Released": "TRUE",
    "UUID": "uuid_2",
    "Version": "version_2",
  },
]
`;

exports[`ModelInfosDocToCsvTransform should correctly transform ModelInfo data to CSV 2`] = `
"UUID,Name,Locale,Description,Version,Released,Previous,ReleaseNotes,OriginURI
uuid_2,"name_2_^°!""§$%&/(",shortCode_2,"description_2_^°!""§$%&/(",version_2,TRUE,previousUUID_2,"releaseNotes_2_^°!""§$%&/(",
"
`;
